### PR TITLE
Implement basic lexer

### PR DESCRIPTION
**Pull request by albertoursino**
Introduction 🤗
---
I have implemented the lexer following pages 8-9-10 of the book.

> The **lexer** breaks up the source code into a list of tokens. Tokens are the smallest syntactic units of a program; they include delimiters, arithmetic symbols, keywords, and identifiers. If a program is like a book, tokens are like individual words.

Implementation Details 📝
---
- **Algorithm**: I followed the pseudocode from Listing 1-3 (page 9);
- **Token Matching**: I used Regular expressions to match tokens, based on the patterns specified in Table 1-1 (page 9).

Testing 🔎
---
Tests from the test suit ✅

```Python
c-compiler$ ./writing-a-c-compiler-tests/test_compiler compiler/compiler_driver.py --chapter 1 --stage lex
----------------------------------------------------------------------
Ran 24 tests in 3.964s

OK
```

Custom tests (https://github.com/d-u-d-e/c-compiler/issues/5) ✅

```Python
c-compiler$ python3 tests/lexer/test_lexer.py 
......
----------------------------------------------------------------------
Ran 6 tests in 0.001s

OK
```

### CODE REVIEW

**Comment by albertoursino** [comment]
This PR contains commits that should not be here. I'm working on it...

**Review by d-u-d-e — APPROVED** [review]
Cool, in a future PR we can move the Token class in a separate file.

**Comment by d-u-d-e** [comment]
Can we get rid of `__init__.py` inside the tests/lexer folder?

**Review by Fabio-Marangoni — APPROVED** [review]

**Comment by Fabio-Marangoni** [comment]
> I get the following error when I try to run the lexer tests from the author's test suite: `ModuleNotFoundError: No module named 'compiler'`. Does that happen to you? That also happens if you run `python compiler/compiler_driver.py` from the project directory.

@d-u-d-e It works fine for me.

**Comment by albertoursino** [comment]
Make sure to do `poetry install`. I fixed this issue in https://github.com/d-u-d-e/c-compiler/pull/6/commits/28857cdb2f7cca7066b1b1c459eca5058ab9eced.

File: compiler/compiler_driver.py [hunk 1]
```diff
@@ -105,8 +111,7 @@ def run_compiler_stages(input_file: str, stage: str) -> None:
         # Execute sequentially the stages
         for current_stage in stages:
             if current_stage == "lex":
-                # TODO: implement the lexer stage
-                raise NotImplementedError("Lexer stage is not implemented.")
+                lexer.run(input_file)
```
Comment (d-u-d-e) [inline compiler/compiler_driver.py:114]: it should be `preprocessed_file.name`

**Review by d-u-d-e — APPROVED** [review]

**Review by CremaLuca — APPROVED** [review]

**Pull request merged**